music21/opensheetmusicdisplay/osmd.py

```diff
@@ -0,0 +1,189 @@
+                // OSMD script has a 'define' call which conflicts with requirejs
+                var _define = window.define // save the define object 
+                window.define = undefined // now the loaded script will ignore requirejs
+                var s = document.createElement( 'script' );
+                // alternative to local file is using a CDN
+                // s.setAttribute( 'src', "https://cdn.jsdelivr.net/npm/opensheetmusicdisplay@0.3.1/build/opensheetmusicdisplay.min.js" );
+                //s.setAttribute( 'src', "{{script_dir}}" );
+                function oncompleted(){
+                    window.define = _define
+                    console.log("loaded OSMD for the first time",opensheetmusicdisplay)
+                    resolve(opensheetmusicdisplay);
+                };
+                {{script_command}}
+                
+            }) 
+        }
+        loadOSMD().then((OSMD)=>{
+            console.log("loaded OSMD",OSMD)
+            var div_id = "{{DIV_ID}}";
+                console.log(div_id)
+            document.querySelector('#'+div_id).innerHTML = "";
+            window.openSheetMusicDisplay = new OSMD.OpenSheetMusicDisplay(div_id);
+            openSheetMusicDisplay
+                .load({{data}})
+                .then(
+                  function() {
+                    console.log("rendering data")
+                    openSheetMusicDisplay.render();
+                  }
+                );
+        })
+        """ \
+        .replace('{{DIV_ID}}',div_id) \
+        .replace('{{data}}',json.dumps(xml)) \
+        .replace('{{script_dir}}',pathname2url(script_dir))
+
+        if offline is True:
+            script_content = open(script_dir).read()
+            script = script.replace('{{script_command}}',"""
+                s.type = 'text/javascript';
+                s.text="""+json.dumps(script_content)+""";
+                document.body.appendChild( s ); // browser will try to load the new script tag
+                oncompleted();
+                """)
+        else:
+            script = script.replace('{{script_command}}',"""
+                s.setAttribute( 'src', 'https://cdn.jsdelivr.net/npm/opensheetmusicdisplay@0.3.1/build/opensheetmusicdisplay.min.js' );
+                s.onload=oncompleted;
+                document.body.appendChild( s ); // browser will try to load the new script tag
+                """)
+
+        return script
+
+    @staticmethod
+    def addDefaultPartName(score):
+        # If no partName is present in the first instrument, OSMD will display the ugly 'partId'
+        allInstruments = list(score.getInstruments(returnDefault=False, recurse=True))
+        
+        if len(allInstruments)==0:
+            # print("adding default inst")
+            defaultInstrument = Piano()
+            defaultInstrument.instrumentName = 'Default'
+            score.insert(None, defaultInstrument)
+            score.coreElementsChanged()
```

Comment:
not needed.

Reply:
Unfortunately this is still needed (thought it would be fixed in the most recent osmd 0.5.1). 

This can be seen in many of the folk corpus pieces such as 'ryansMammoth/NorfolkHornpipe.abc', which will convert to musicXML with part-name `<part-name />` in them. This will then display in the score as 'Instr. P6d2a255fb4b93a88d140fb66e502d885' since osmd uses the id when part-name is missing.

That being said my attempted fix doesn't seem to work! Even changing `partName` and calling `coreElementsChanged` willl not reflect in the piece when written to musicXML.

Reply:
An example of what I mean:
![image](https://user-images.githubusercontent.com/11251397/46961820-88e3bc00-d099-11e8-992b-6a866c8751ec.png)
![image](https://user-images.githubusercontent.com/11251397/46961863-a31d9a00-d099-11e8-94ab-a72465a6350c.png)